{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the rna string: AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\n",
      "MAMAPRTEINSTRING\n"
     ]
    }
   ],
   "source": [
    "CodonTable = {\n",
    "    'UUU': 'F',     'CUU': 'L',     'AUU': 'I',     'GUU': 'V',\n",
    "    'UUC': 'F',     'CUC': 'L',     'AUC': 'I',     'GUC': 'V',\n",
    "    'UUA': 'L',     'CUA': 'L',     'AUA': 'I',     'GUA': 'V',\n",
    "    'UUG': 'L',     'CUG': 'L',     'AUG': 'M',     'GUG': 'V',\n",
    "    'UCU': 'S',     'CCU': 'P',     'ACU': 'T',     'GCU': 'A',\n",
    "    'UCC': 'S',     'CCC': 'P',     'ACC': 'T',     'GCC': 'A',\n",
    "    'UCA': 'S',     'CCA': 'P',     'ACA': 'T',     'GCA': 'A',\n",
    "    'UCG': 'S',     'CCG': 'P',     'ACG': 'T',     'GCG': 'A',\n",
    "    'UAU': 'Y',     'CAU': 'H',     'AAU': 'N',     'GAU': 'D',\n",
    "    'UAC': 'Y',     'CAC': 'H',     'AAC': 'N',     'GAC': 'D',\n",
    "    'UAA': 'Stop',  'CAA': 'Q',     'AAA': 'K',     'GAA': 'E',\n",
    "    'UAG': 'Stop',  'CAG': 'Q',     'AAG': 'K',     'GAG': 'E',\n",
    "    'UGU': 'C',     'CGU': 'R',     'AGU': 'S',     'GGU': 'G',\n",
    "    'UGC': 'C',     'CGC': 'R',     'AGC': 'S',     'GGC': 'G',\n",
    "    'UGA': 'Stop',  'CGA': 'R',     'AGA': 'R',     'GGA': 'G',\n",
    "    'UGG': 'W',     'CGG': 'R',     'AGG': 'R',     'GGG': 'G'\n",
    "}\n",
    "\n",
    "string = input(\"Enter the rna string: \")\n",
    "pr = ''\n",
    "\n",
    "for i in range(0,len(string),3):\n",
    "    s = CodonTable[string[i:i+3]]\n",
    "    if s == 'Stop':\n",
    "        break\n",
    "    pr += s\n",
    "    \n",
    "print(pr)\n",
    "     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the dna string: ATGGCCATGGCCCCCAGAACTGAGATCAATAGTACCCGTATTAACGGGTGA\n",
      "Enter the amino acid string: MA\n",
      "ATGGCC\n",
      "GGCCAT\n",
      "ATGGCC\n"
     ]
    }
   ],
   "source": [
    "CodonTable = {\n",
    "    'UUU': 'F',     'CUU': 'L',     'AUU': 'I',     'GUU': 'V',\n",
    "    'UUC': 'F',     'CUC': 'L',     'AUC': 'I',     'GUC': 'V',\n",
    "    'UUA': 'L',     'CUA': 'L',     'AUA': 'I',     'GUA': 'V',\n",
    "    'UUG': 'L',     'CUG': 'L',     'AUG': 'M',     'GUG': 'V',\n",
    "    'UCU': 'S',     'CCU': 'P',     'ACU': 'T',     'GCU': 'A',\n",
    "    'UCC': 'S',     'CCC': 'P',     'ACC': 'T',     'GCC': 'A',\n",
    "    'UCA': 'S',     'CCA': 'P',     'ACA': 'T',     'GCA': 'A',\n",
    "    'UCG': 'S',     'CCG': 'P',     'ACG': 'T',     'GCG': 'A',\n",
    "    'UAU': 'Y',     'CAU': 'H',     'AAU': 'N',     'GAU': 'D',\n",
    "    'UAC': 'Y',     'CAC': 'H',     'AAC': 'N',     'GAC': 'D',\n",
    "    'UAA': 'Stop',  'CAA': 'Q',     'AAA': 'K',     'GAA': 'E',\n",
    "    'UAG': 'Stop',  'CAG': 'Q',     'AAG': 'K',     'GAG': 'E',\n",
    "    'UGU': 'C',     'CGU': 'R',     'AGU': 'S',     'GGU': 'G',\n",
    "    'UGC': 'C',     'CGC': 'R',     'AGC': 'S',     'GGC': 'G',\n",
    "    'UGA': 'Stop',  'CGA': 'R',     'AGA': 'R',     'GGA': 'G',\n",
    "    'UGG': 'W',     'CGG': 'R',     'AGG': 'R',     'GGG': 'G'\n",
    "}\n",
    "\n",
    "def revDna(dna):\n",
    "    complement = {'A':'T' , 'C':'G' , 'T':'A' , 'G':'C'}   \n",
    "    rc = ''\n",
    "\n",
    "    for letter in dna:\n",
    "        rc += complement[letter]\n",
    "        \n",
    "    rc = \"\".join(reversed(rc))\n",
    "    return rc\n",
    "\n",
    "def translation(rna):\n",
    "    pr = ''\n",
    "    \n",
    "    for i in range(0,len(rna),3):\n",
    "        s = CodonTable[rna[i:i+3]]\n",
    "        if s == 'Stop':\n",
    "            break\n",
    "        pr += s\n",
    "    \n",
    "    return pr\n",
    "\n",
    "def transcribe(dna):\n",
    "    return dna.replace('T', 'U') \n",
    "\n",
    "\n",
    "dna = input(\"Enter the dna string: \")\n",
    "peptide = input(\"Enter the amino acid string: \")\n",
    "\n",
    "p = len(peptide)*3\n",
    "\n",
    "s1 = []\n",
    "s2 = []\n",
    "\n",
    "for i in range(len(dna)-p+1):\n",
    "    s1.append(dna[i:i+p])\n",
    "    s2.append(revDna(dna[i:i+p]))\n",
    "\n",
    "for i in range(len(s1)):\n",
    "    t1 = transcribe(s1[i])\n",
    "    t2 = transcribe(s2[i])\n",
    "    r1 = translation(t1)\n",
    "    r2 = translation(t2)\n",
    "    \n",
    "    if r1 == peptide or r2==peptide:\n",
    "        print(s1[i])\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter the peptide string: LEQN\n",
      "0 113 114 128 129 227 242 242 257 355 356 370 371 484 \n"
     ]
    }
   ],
   "source": [
    "mass = {}\n",
    "mass['G'] = 57\n",
    "mass['A'] = 71\n",
    "mass['S'] = 87\n",
    "mass['P'] = 97\n",
    "mass['V'] = 99\n",
    "mass['T'] = 101\n",
    "mass['C'] = 103\n",
    "mass['I'] = 113\n",
    "mass['L'] = 113\n",
    "mass['N'] = 114\n",
    "mass['D'] = 115\n",
    "mass['K'] = 128\n",
    "mass['Q'] = 128\n",
    "mass['E'] = 129\n",
    "mass['M'] = 131\n",
    "mass['H'] = 137\n",
    "mass['F'] = 147\n",
    "mass['R'] = 156\n",
    "mass['Y'] = 163\n",
    "mass['W'] = 186\n",
    "\n",
    "\n",
    "def total_Mass(peptide):\n",
    "    total = 0\n",
    "    for ch in peptide:\n",
    "        total += mass[ch]\n",
    "        \n",
    "    return total\n",
    "\n",
    "peptide = input(\"Enter the peptide string: \")\n",
    "\n",
    "pl = len(peptide)\n",
    "res = [0]\n",
    "\n",
    "for x in range(1,pl):\n",
    "    for i in range(pl):\n",
    "        if i+x > len(peptide):\n",
    "            y = i+x-len(peptide)\n",
    "            res.append(total_Mass(peptide[i:]+peptide[:y]))\n",
    "        else:\n",
    "            res.append(total_Mass(peptide[i:i+x]))\n",
    "            \n",
    "res.append(total_Mass(peptide))\n",
    "res.sort()\n",
    "\n",
    "ans = \"\"\n",
    "\n",
    "for x in res:\n",
    "    ans +=  str(x) + \" \"\n",
    "    \n",
    "print(ans)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task-6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "CTTA\n",
      "ACCA\n",
      "TACC\n",
      "GGCT\n",
      "GCTT\n",
      "TTAC\n",
      "quit\n",
      "GGCTTACCA\n"
     ]
    }
   ],
   "source": [
    "def debruijn(patterns):\n",
    "    kmers = []\n",
    "    for pattern in patterns:\n",
    "        kmers = kmers + composition(len(pattern), pattern, uniq=True)\n",
    "        \n",
    "    kmers = set(kmers)\n",
    "    res = {}\n",
    "    \n",
    "    for k in kmers:\n",
    "        res[k] = []\n",
    "        \n",
    "    for kmer in patterns:\n",
    "        res[prefix(kmer)].append(suffix(kmer))\n",
    "        \n",
    "    return res\n",
    "\n",
    "\n",
    "def GenomePath(reads):\n",
    "    path = reads[0]\n",
    "    for e in reads[1:]:\n",
    "        path += e[len(e)-1:]\n",
    "\n",
    "    return path\n",
    "\n",
    "def eulerPath(dict):\n",
    "    stack = []\n",
    "    count = balance(dict)   \n",
    "    stack.append([k for k, v in count.items() if v==-1][0])\n",
    "    path = []\n",
    "    \n",
    "    while stack != []:\n",
    "        u = stack[-1]        \n",
    "        try:\n",
    "            v = dict[u][0]\n",
    "            stack.append(v)\n",
    "            dict[u].remove(v)\n",
    "        except:           \n",
    "            path.append(stack.pop())\n",
    "        \n",
    "    return path[::-1]\n",
    "\n",
    "\n",
    "def composition(k, text, uniq=False):\n",
    "    kmers = []\n",
    "    for i in range(len(text)+1-k):\n",
    "        kmers.append(text[i:i+k-1])\n",
    "    if uniq:\n",
    "        return sorted(list(kmers))\n",
    "    else:\n",
    "        return sorted(kmers)\n",
    "\n",
    "def balance(adj):\n",
    "    count = dict.fromkeys(adj.keys(), 0)\n",
    "    for node in adj.keys():\n",
    "        for out in adj[node]:\n",
    "            count[node] -= 1\n",
    "            try:\n",
    "                count[out] += 1\n",
    "            except:\n",
    "                count[out] = 1\n",
    "    return count\n",
    "\n",
    "\n",
    "def suffix(string):\n",
    "    return string[1:]\n",
    "\n",
    "\n",
    "def prefix(string):\n",
    "    return string[0:-1]\n",
    "\n",
    "k = int(input())\n",
    "dna = []\n",
    "a = ''\n",
    "while True:\n",
    "    a = input()\n",
    "    if a == 'quit':\n",
    "        break\n",
    "    dna.append(a)\n",
    "\n",
    "ans = GenomePath(eulerPath(debruijn(dna)))\n",
    "print(ans)\n",
    "    \n",
    "#4\n",
    "#CTTA\n",
    "#ACCA\n",
    "#TACC\n",
    "#GGCT\n",
    "#GCTT\n",
    "#TTAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
